File: 강의 자료/02-알고리즘/12-트리/문제별 코드/1068-트리/solution.cpp
#include <iostream>
#include <algorithm>
#pragma warning(disable:4996)
using namespace std;
#define NM 55

#include <vector>
int n, root, erased, leaf[NM];
vector<int> child[NM];

void input() {
	cin >> n;
	for (int i = 0; i < n; i++) {
		int par;
		cin >> par;
		if (par == -1) {
			root = i;
		}
		else {
			child[par].push_back(i);
		}
	}
	cin >> erased;
}

void DFS(int x, int par) {
	if (child[x].empty())
		leaf[x]++;
	for (int y : child[x]) {
		if (y == par) continue;
		DFS(y, x);
		leaf[x] += leaf[y];
	}
}

void pro() {
	for (int i = 0; i < n; i++) {
		for (auto it = child[i].begin(); it != child[i].end(); it++){
			if (*it == erased) {
				child[i].erase(it);
				break;
			}
		}
	}

	if (root != erased) DFS(root, -1);

	cout << leaf[root];
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	input();
	pro();
	return 0;
}

File: 강의 자료/02-알고리즘/12-트리/문제별 코드/1068-트리/solution.py
import sys
from collections import deque
si = sys.stdin.readline
n = int(si())
parent = list(map(int, si().split()))
child = [[] for _ in range(n)]
leaf = [0] * n
erased = int(si())
root = 0
for i in range(n):
    if parent[i] == -1:
        root = i
    else:
        if erased == i:
            continue
        child[parent[i]].append(i)

def dfs(x, par):
    if not child[x]:
        leaf[x] = 1
    for y in child[x]:
        if y == par:
            continue
        dfs(y, x)
        leaf[x] += leaf[y]

if root != erased:
    dfs(root, -1)
print(leaf[root])